internal/overlord/cmdstate/cmdmgr.go

```diff
@@ -0,0 +1,654 @@
+		ttys = append(ttys, stdinWriter)
+		go func() {
+			<-wsutil.WebsocketRecvStream(stdinWriter, ioConn)
+			stdinWriter.Close()
+		}()
+
+		// Start goroutine to receive from cmd.Stdout pipe and write to "io"
+		// websocket.
+		stdoutReader, stdoutWriter, err := os.Pipe()
+		if err != nil {
+			return err
+		}
+		ptys = append(ptys, stdoutReader)
+		ttys = append(ttys, stdoutWriter)
+		stdout = stdoutWriter
+		stderr = stdoutWriter // stderr will be overwritten below if splitStderr true
+		wg.Add(1)
+		go func() {
+			<-wsutil.WebsocketSendStream(ioConn, stdoutReader, -1)
+			stdoutReader.Close()
+			wg.Done()
+		}()
+	}
+
+	if e.splitStderr {
+		// Start goroutine to receive from cmd.Stderr pipe and write to a
+		// separate "stderr" websocket.
+		stderrReader, stderrWriter, err := os.Pipe()
+		if err != nil {
+			return err
+		}
+		ptys = append(ptys, stderrReader)
+		ttys = append(ttys, stderrWriter)
+		stderr = stderrWriter
+		stderrConn := e.getWebsocket(wsStderr)
+		wg.Add(1)
+		go func() {
+			<-wsutil.WebsocketSendStream(stderrConn, stderrReader, -1)
+			stderrReader.Close()
+			wg.Done()
+		}()
+	}
+
+	if e.timeout != 0 {
+		var cancel func()
+		ctx, cancel = context.WithTimeout(ctx, e.timeout)
+		defer cancel()
+	}
+
+	cmd := exec.CommandContext(ctx, e.command[0], e.command[1:]...)
+
+	for k, v := range e.env {
+		cmd.Env = append(cmd.Env, fmt.Sprintf("%s=%s", k, v))
+	}
+
+	cmd.Dir = e.workingDir
+
+	cmd.Stdin = stdin
+	cmd.Stdout = stdout
+	cmd.Stderr = stderr
+
+	cmd.SysProcAttr = &syscall.SysProcAttr{}
+	if e.uid != nil && e.gid != nil {
+		cmd.SysProcAttr.Credential = &syscall.Credential{
+			Uid: uint32(*e.uid),
+			Gid: uint32(*e.gid),
+		}
+	}
+
+	// Creates a new session if the calling process is not a process group
+	// leader. The calling process is the leader of the new session, the
+	// process group leader of the new process group, and has no controlling
+	// terminal. This is important to allow remote shells to handle Ctrl+C.
+	cmd.SysProcAttr.Setsid = true
+
+	// Make the given terminal the controlling terminal of the calling
+	// process. The calling process must be a session leader and not have a
+	// controlling terminal already. This is important as allows Ctrl+C to
+	// work as expected for non-shell programs.
+	if e.useTerminal {
+		cmd.SysProcAttr.Setctty = true
+	}
+
+	finisher := func(exitCode int, cmdErr error) error {
+		for _, tty := range ttys {
+			_ = tty.Close()
+		}
+
+		// Close the control channel, if connected.
+		controlConn := e.getWebsocket(wsControl)
+		if controlConn != nil {
+			_ = controlConn.Close()
+		}
+
+		close(childDead)
+
+		wg.Wait()
+
+		for _, pty := range ptys {
+			_ = pty.Close()
+		}
+
+		setApiData(change, exitCode)
+
+		return cmdErr
+	}
+
+	// Start the command!
+	err = cmd.Start()
+	if err != nil {
+		return finisher(-1, err)
+	}
+
+	// Send its PID to the control loop.
+	pidCh <- cmd.Process.Pid
+
+	// Wait for it to finish.
+	err = cmd.Wait()
+
```

Comment:
Every path from here on calls finisher. Wouldn't it make the logic easier to read/write if the body of finisher is moved here instead, so that the logic becomes linear again?  If I read it correctly, the function doesn't even look into the provided arguments until the last two lines, so everything else is just unconditionally run.

Reply:
Done. The only slight annoyance is that we also want to do the finishing logic after `cmd.Start()`, if that has an error. But I've updated in a way that makes it linear and I think more obvious.